core/src/main/java/com/linecorp/armeria/server/HttpResponseSubscriber.java

```diff
@@ -226,6 +219,14 @@ public void onNext(HttpObject o) {
                 return;
         }
 
+        if (endOfStream && !(o instanceof HttpHeaders)) {
+            final HttpHeaders additionalTrailers = reqCtx.additionalResponseTrailers();
+            if (!additionalTrailers.isEmpty()) {
```

Comment:
Perhaps this part could be handled better in the `case NEEDS_DATA_OR_TRAILING_HEADERS:` part?

```java
case NEEDS_DATA_OR_TRAILING_HEADERS: {
    if (o instanceof HttpHeaders) {
        ...
    } else if (endOfStream) { // Last DATA frame
        final HttpHeaders additionalTrailers = reqCtx.additionalResponseTrailers();
        if (!additionalTrailers.isEmpty()) {
            write(o, false);
            o = additionalTrailers;
        }
    }
    break;
}
```

Reply:
Ah yeah